prql-compiler/prqlc/src/cli.rs

```diff
@@ -128,46 +154,55 @@ impl Cli {
                 // combine with source
                 combine_prql_and_frames(source, frames).as_bytes().to_vec()
             }
-            Cli::Resolve(_) => {
+            Commands::Resolve(_) => {
                 let ast = prql_to_pl(source)?;
                 let ir = semantic::resolve(ast)?;
-
                 serde_json::to_string_pretty(&ir)?.into_bytes()
+                // match args.format {
+                //     Some(Format::Json) | None => serde_json::to_string_pretty(&ir)?.into_bytes(),
+                //     Some(Format::Yaml) => anyhow::bail!("YAML output is not yet supported for PL"),
+                //     // Some(Format::Yaml) => serde_yaml::to_string(&ir)?.into_bytes(),
+                // }
             }
-            Cli::Compile(_) => compile(source, &Options::default())?.as_bytes().to_vec(),
-            Cli::Watch(_) => unreachable!(),
+            // TODO: Allow passing the `Options` to the CLI; map those through.
+            // We already do this in Watch.
+            Commands::Compile(_) => compile(source, &Options::default())?.as_bytes().to_vec(),
+            Commands::Watch(_) => unreachable!(),
         })
     }
 
     fn read_input(&mut self) -> Result<(String, String)> {
-        use Cli::*;
-        match self {
-            Parse(io) | Format(io) | Debug(io) | Annotate(io) | Resolve(io) | Compile(io) => {
-                // Don't wait without a prompt when running `prql-compiler compile` —
-                // it's confusing whether it's waiting for input or not. This
-                // offers the prompt.
-                if is_stdin(&io.input) && atty::is(atty::Stream::Stdin) {
-                    println!("Enter PRQL, then ctrl-d:");
-                    println!();
-                }
-
-                let mut source = String::new();
-                io.input.read_to_string(&mut source)?;
-                let source_id = (*io.input.path()).to_str().unwrap().to_string();
-                Ok((source, source_id))
-            }
-            Cli::Watch(_) => unreachable!(),
+        use Commands::*;
+        let mut input = match self {
+            // TODO: are these clones sound?
```

Comment:
https://docs.rs/clio/0.2.7/clio/enum.Input.html#impl-Clone-for-Input

It looks like it is sound. It should not have any performance implications.

Reply:
Yeah, in the narrow "does it compile without the clones" — it doesn't. But I was a bit confused by this — in the old approach they weren't needed. 

I think this is one of those times that the compiler was being helpful, we adjust the flow of the code, the compiler can no longer see that it can still be helpful, and so the result is somewhat confusing...

Reply:
I left a TODO about inverting how this function is called; I think that is likely to make the code better and avoid this problem...